File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        vector<vector<int>> res;
        queue<pair<TreeNode*, int>> que;
        que.push({root, 0});
        vector<int> row;
        int cur_level = 0;
        while (!que.empty()) {
            auto [node, level] = que.front();
            que.pop();
            if (level != cur_level) {
                res.push_back(row);
                cur_level = level;
                row = {};
            }
            row.push_back(node->val);
            if (node->left) que.push({node->left, level + 1});
            if (node->right) que.push({node->right, level + 1});
        }

        if (!row.empty()) res.push_back(row);
        return res;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = [0]
        def dfs(node):
            if not node:
                return
            
            l = dfs(node.left) if node.left else -1
            r = dfs(node.right) if node.right else -1
            res[0] = max(res[0], l + 2 + r)
            return max(l, r) + 1

        dfs(root)
        return res[0]